datafusion/common/src/tree_node.rs

```diff
@@ -769,6 +770,263 @@ impl<T> Transformed<T> {
     }
 }
 
+/// [`Container`] contains elements that a function can be applied on or mapped. The
+/// elements of the container are siblings so the continuation rules are similar to
+/// [`TreeNodeRecursion::visit_sibling`] / [`Transformed::transform_sibling`].
+pub trait Container<'a, T: 'a>: Sized {
+    fn apply_elements<F: FnMut(&'a T) -> Result<TreeNodeRecursion>>(
+        &'a self,
+        f: F,
+    ) -> Result<TreeNodeRecursion>;
+
+    fn map_elements<F: FnMut(T) -> Result<Transformed<T>>>(
+        self,
+        f: F,
+    ) -> Result<Transformed<Self>>;
+}
+
+impl<'a, T: 'a, C: Container<'a, T>> Container<'a, T> for Box<C> {
+    fn apply_elements<F: FnMut(&'a T) -> Result<TreeNodeRecursion>>(
```

Comment:
Can we somehow simplify these declarations by aliasing like
`type ApplyFn<'a, T> = dyn FnMut(&'a T) -> Result<TreeNodeRecursion>;`
`type MapFn<T> = dyn FnMut(T) -> Result<Transformed<T>>;`

Reply:
Good idea, I will fix it soon.

Reply:
Actually, introducing `dyn FnMut` (trait objects) would mean dynamic dispatch and so come with performance penalty. Trait aliase (https://doc.rust-lang.org/beta/unstable-book/language-features/trait-alias.html) is an unstalbe feature yet.